include/aws/crt/ScopedRWLock.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include <aws/common/rw_lock.h>
+
+namespace Aws
+{
+    namespace Crt
+    {
+        /**
+         * Custom implementation of an ScopedReadLock type. Wrapping the aws_rw_lock.
+         * On creation, the ScopedTryReadLock will attempts to acquire the lock and returns immediately if it can not.
+         * The lock will be unlocked on destruction.
+         * Use aws_last_error() or operator bool() to check if the lock get acquired successfully.
+         */
+
+        class ScopedTryReadLock
```

Comment:
Also, why are we using try and not just lock?

Reply:
Try lock will immediately return if it failed to acquire the lock. If we are in termination, then we would like to directly failed the operation instead of waiting on the lock.

Reply:
I setup in the way that the write lock will lock once `Close()` is invoked and only get unlock/clean up in destructor. In that case, we can fail all operations after `Close()` invoked.

Reply:
The write lock would only be held for as long as it take to dec ref the native client and change the pointer.  I don't see why we need to hold it arbitrarily long until the destructor.

Reply:
That makes sense. Similarly, I will make a scopedWriteLock for the write lock, so it would unlock once it exit the `Close()` function call. 
Also, I will keep the **Try**ReadLock if no further comments on it.